File: app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionAdapter.java
package org.dhis2.usescases.programStageSelection;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.dhis2.R;
import org.dhis2.databinding.ItemProgramStageBinding;
import org.hisp.dhis.android.core.program.ProgramStage;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Created by Cristian on 13/02/2018.
 */

public class ProgramStageSelectionAdapter extends RecyclerView.Adapter<ProgramStageSelectionViewHolder> {

    private ProgramStageSelectionContract.Presenter presenter;
    private List<ProgramStage> programStages;

    ProgramStageSelectionAdapter(@NonNull ProgramStageSelectionContract.Presenter presenter) {
        this.presenter = presenter;
    }

    public void setProgramStages(List<ProgramStage> programStages) {
        this.programStages = programStages;
    }

    @Override
    public ProgramStageSelectionViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ItemProgramStageBinding binding = DataBindingUtil.inflate(inflater, R.layout.item_program_stage, parent, false);
        return new ProgramStageSelectionViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ProgramStageSelectionViewHolder holder, int position) {
        ProgramStage programStage = programStages.get(position);
        holder.bind(presenter, programStage);
    }

    @Override
    public int getItemCount() {
        return programStages != null ? programStages.size() : 0;
    }
}

File: app/src/main/java/org/dhis2/usescases/programStageSelection/ProgramStageSelectionAdapter.kt
package org.dhis2.usescases.programStageSelection

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import org.dhis2.R
import org.dhis2.databinding.ItemProgramStageBinding
import org.hisp.dhis.android.core.program.ProgramStage

class ProgramStageSelectionAdapter(
    val onItemClick: (ProgramStage) -> Unit,
) : ListAdapter<ProgramStageData, ProgramStageSelectionViewHolder>(object :
    DiffUtil.ItemCallback<ProgramStageData>() {
    override fun areItemsTheSame(oldItem: ProgramStageData, newItem: ProgramStageData): Boolean {
        return oldItem.programStage.uid() == newItem.programStage.uid()
    }

    override fun areContentsTheSame(oldItem: ProgramStageData, newItem: ProgramStageData): Boolean {
        return oldItem == newItem
    }
}) {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ProgramStageSelectionViewHolder {
        val binding: ItemProgramStageBinding = DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.item_program_stage,
            parent,
            false,
        )
        return ProgramStageSelectionViewHolder(binding, onItemClick)
    }

    override fun onBindViewHolder(holder: ProgramStageSelectionViewHolder, position: Int) {
        holder.bind(getItem(position))
    }
}
